### PR TITLE
Plugin Traffic Dump

### PR DESCRIPTION
**Pull request by dyrock**
Plugin Traffic Dump captures traffic on a per session basis. A sampling ratio can be set via plugin.config or traffic_ctl to dump one out of n sessions. The dump file schema can be found within this PR https://github.com/dyrock/trafficserver/blob/5eee7dec52ebb4978afe0810092eb0baf97fb284/tests/tools/lib/replay_schema.json

### CODE REVIEW

**Comment by maskit** [comment]
Just out of curiosity, does it handle HTTP/2 Server Push?

It might be nice if the log has a field for request origin (e.g. client, plugin or internal).

**Comment by dyrock** [comment]
@maskit We have `TSHttpTxnIsInternal` but no api for retrieving plugin origin (like plugin tag/id). Maybe we should add an api for that and call it after `TSHttpTxnIsInternal`?

**Comment by maskit** [comment]
> Also, I think it does handle HTTP2 Server Push.

It'll record the transaction but how do we know if it was server push? That was my real question :)

**Comment by dyrock** [comment]
> Also, I think it does handle HTTP2 Server Push.
@maskit I was wrong about this. :/ The plugin records h2 transactions but simply ignores H2 server push for now. (We don't actually know what to do with them but you are welcome to throw out suggestions/proposals)

File: tests/tools/lib/replay_schema.json [hunk 1]
```diff
@@ -0,0 +1,209 @@
+{
```
Comment (maskit) [inline tests/tools/lib/replay_schema.json]: Can we put AL2 header here?

File: plugins/experimental/traffic_dump/traffic_dump.cc [hunk 2]
```diff
@@ -0,0 +1,592 @@
...
+{
+  std::string result = "{";
+  int len            = 0;
+  const char *cp     = nullptr;
+  TSMLoc url_loc     = nullptr;
+
+  // Log scheme+method or status+reason based on header type
+  if (TSHttpHdrTypeGet(buffer, hdr_loc) == TS_HTTP_TYPE_REQUEST) {
+    // 1. "version"
+    int version = TSHttpHdrVersionGet(buffer, hdr_loc);
```
Comment (shinrich) [inline plugins/experimental/traffic_dump/traffic_dump.cc]: I don't think this will capture HTTP/2.

File: tests/tools/lib/replay_schema.json (hunk 1 above)
Comment (shinrich) [reply tests/tools/lib/replay_schema.json]: What is an AL2 header?

File: plugins/experimental/traffic_dump/traffic_dump.cc [hunk 3]
```diff
@@ -0,0 +1,592 @@
...
+    // Create new per session data
+    SsnData *ssnData = new SsnData;
+    TSHttpSsnArgSet(ssnp, s_arg_idx, ssnData);
+
+    TSContDataSet(ssnData->aio_cont, ssnData);
+
+    // 1. "protocol":(string)
+    const char *protocol[10];
+    int count = 0;
+    TSAssert(TS_SUCCESS == TSHttpSsnClientProtocolStackGet(ssnp, 10, protocol, &count));
```
Comment (dyrock) [inline plugins/experimental/traffic_dump/traffic_dump.cc]: H2 would be captured here as `protocol` property of session object.

File: tests/tools/lib/replay_schema.json (hunk 1 above)
Comment (maskit) [reply tests/tools/lib/replay_schema.json]: Apache License 2

File: plugins/experimental/traffic_dump/traffic_dump.cc (hunk 2 above)
Comment (shinrich) [reply plugins/experimental/traffic_dump/traffic_dump.cc]: @dyrock clarified that the H1/H2 information is captured at the session level which seems appropriate.

File: tests/tools/lib/replay_schema.json (hunk 1 above)
Comment (dyrock) [reply tests/tools/lib/replay_schema.json]: On it!

Comment (SolidWallOfCode) [reply tests/tools/lib/replay_schema.json]: I think the best option is to add to the `description` field the text "Licensed under Apache V2 https://www.apache.org/licenses/LICENSE-2.0".

**Comment by SolidWallOfCode** [comment]
Looks good, just need the tweak for the license on the schema file. At some point we will want to do a pass for better performance, once some of the core utilities are available to plugins.

**Comment by bryancall** [comment]
[approve ci autest]

**Comment by d2r** [comment]
[approve ci autest]

**Comment by SolidWallOfCode** [comment]
[approve ci autest]

**Pull request merged**